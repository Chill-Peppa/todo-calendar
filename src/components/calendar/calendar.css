.calendar {
  background-color: #fff;
  max-width: 500px;
  border-radius: 15px;
  padding: 40px 40px;
  box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
}

.calendar__header {
  display: flex;
  justify-content: space-between;
}

.calendar__title {
  font-weight: 600;
}

.calendar__button-container {
  display: flex;
  align-items: center;
  gap: 20px;
}

.calendar__button {
  display: flex;
  align-items: center;
  border: none;
  background: transparent;
  cursor: pointer;
}

.calendar__next {
  height: 25px;
  width: 25px;
  opacity: 1;
}

.calendar__next:hover {
  opacity: 0.65;
  transition: all 1s linear;
}

.calendar__prev {
  height: 25px;
  width: 25px;
  opacity: 1;
}

.calendar__prev:hover {
  opacity: 0.7;
  transition: all 1s linear;
}

.calendar__weekdays {
  margin: 35px 0 20px 0;
  list-style: none;
  font-weight: 600;
  display: flex;
  justify-content: space-between;
}

@media screen and (max-width: 500px) {
  .calendar__weekdays {
    width: 97%;
    align-items: center;
  }
}

.calendar__days {
  list-style: none;
  font-weight: 500;
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  grid-template-rows: repeat(5, 1fr);
  gap: 20px 38px;
}

@media screen and (max-width: 555px) {
  .calendar__days {
    gap: 20px 10px;
  }
}

.calendar__day {
  padding: 2px;
  display: flex;
  justify-content: center;
  /* border-radius: 50%; */
  cursor: pointer;
}

.calendar__day_event {
  border-bottom: 3px solid #a6bdd7;
}

.calendar__day_holiday {
  color: red;
}

.calendar__day_inactive {
  color: #aaa;
}

.calendar__week-deals {
  margin-top: 30px;
  font-size: 20px;
}

@media screen and (max-width: 475px) {
  .calendar__week-deals {
    font-size: 18px;
  }
}

.calendar__see-more {
  cursor: pointer;
  color: #5d9b9b;
  opacity: 1;
}

.calendar__see-more:hover {
  opacity: 0.6;
  transition: opacity 0.7s ease-in-out;
}
