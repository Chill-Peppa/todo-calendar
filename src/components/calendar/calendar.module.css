.calender {
  background-color: #fff;
  width: 500px;
  border-radius: 15px;
  padding: 40px 30px;
  box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
}

.header {
  display: flex;
  justify-content: space-between;
}

.currentData {
  font-weight: 600;
}

.arrowsContainer {
  display: flex;
  align-items: center;
  gap: 20px;
}

.left {
  height: 25px;
  width: 25px;
  cursor: pointer;
  opacity: 1;
}

.left:hover {
  opacity: 0.7;
  transition: all 1s linear;
}

.right {
  composes: left;
}

.weekDays {
  margin: 35px 0 20px 0;
  list-style: none;
  font-weight: 600;
  display: flex;
  justify-content: space-between;
}

.days {
  list-style: none;
  font-weight: 500;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr 1fr 1fr 1fr;
  gap: 20px 37px;
}

.day {
  border: 1px solid #a6bdd7;
  padding: 2px;
  display: flex;
  justify-content: center;
  border-radius: 50%;
  cursor: pointer;
}
