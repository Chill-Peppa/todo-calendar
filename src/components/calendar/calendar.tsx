import React, { useState, useEffect } from 'react';
import './calendar.css';
import left from '../../assets/images/left-arrow.svg';
import right from '../../assets/images/right-arrow.svg';

import { weekDays } from '../../utils/constants';
import { months } from '../../utils/constants';

interface ICalendarProps {
  onOpenTodo: (e: React.MouseEvent<HTMLLIElement>) => void;
  onOpenWeekTasks: () => void;
}

const Calendar: React.FC<ICalendarProps> = ({
  onOpenTodo,
  onOpenWeekTasks,
}) => {
  const date = new Date();
  const [currentMonth, setCurrentMonth] = useState(date.getMonth());
  const [currentYear, setCurrentYear] = useState(date.getFullYear());
  const [daysInMonth, setDaysInMonth] = useState<
    { day: number; isHoliday: number }[]
  >([]);
  const [dataLoaded, setDataLoaded] = useState<boolean>(false);

  useEffect(() => {
    const fetchData = async () => {
      const days = await getAllDaysInMonth();
      setDaysInMonth(days);
      setDataLoaded(true);
    };

    fetchData();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentYear, currentMonth]);

  const returnDate = () => {
    return `${months[currentMonth]} ${currentYear}`;
  };

  const getAllDaysInMonth = async () => {
    let firstDayOfMonth = new Date(currentYear, currentMonth, 1).getDay();
    const lastDayOfMonth = new Date(currentYear, currentMonth + 1, 0).getDate();
    const lastDateOfLastMonth = new Date(
      currentYear,
      currentMonth,
      0,
    ).getDate();

    if (firstDayOfMonth === 0) {
      firstDayOfMonth = 6; // Переназначаем воскресенье на 6 (вместо 0)
    } else {
      firstDayOfMonth--; // Коррекция нумерации остальных дней
    }

    let allDaysInMonthArray: {
      day: number;
      isHoliday: number;
    }[] = [];

    for (let i = firstDayOfMonth; i > 0; i--) {
      allDaysInMonthArray = [
        ...allDaysInMonthArray,
        { day: lastDateOfLastMonth - i + 1, isHoliday: 2 },
      ];
    }

    for (let i = 1; i <= lastDayOfMonth; i++) {
      try {
        const response = await fetch(
          `https://isdayoff.ru/api/getdata?year=${currentYear}&month=${
            currentMonth + 1
          }&day=${i}`,
        );
        if (response.ok) {
          const isHoliday = await response.json();
          allDaysInMonthArray = [
            ...allDaysInMonthArray,
            { day: i, isHoliday: isHoliday },
          ];
        } else {
          throw new Error('Fetch error');
        }
      } catch (error) {
        console.error('Error fetch:', error);
      }
    }

    return allDaysInMonthArray;
  };

  const onNextButtonClick = () => {
    let newMonth = (currentMonth + 1) % 12;
    let newYear = currentYear;
    if (newMonth === 0) {
      newYear++;
    }

    setCurrentMonth(newMonth);
    setCurrentYear(newYear);
    setDataLoaded(false);
  };

  const onPrevButtonClick = () => {
    let newMonth = (currentMonth - 1 + 12) % 12;
    let newYear = currentYear;
    if (currentMonth === 0) {
      newYear--;
    }

    setCurrentMonth(newMonth);
    setCurrentYear(newYear);
    setDataLoaded(false);
  };

  return (
    <section className="calendar">
      <div className="calendar__header">
        {dataLoaded ? (
          <h1 className="calendar__title">{returnDate()}</h1>
        ) : (
          <span>загрузка</span>
        )}
        <div className="calendar__button-container">
          <button onClick={onPrevButtonClick} className="calendar__button">
            <img
              src={left as unknown as string}
              alt="left arrow"
              className="calendar__prev"
            />
          </button>
          <button onClick={onNextButtonClick} className="calendar__button">
            <img
              src={right as unknown as string}
              alt="right arrow"
              className="calendar__next"
            />
          </button>
        </div>
      </div>

      <div className="calendar__main">
        <ul className="calendar__weekdays">
          {weekDays.map((weekDay, index) => (
            <li className="calendar__weekday" key={index}>
              {weekDay}
            </li>
          ))}
        </ul>

        <ul className="calendar__days">
          {daysInMonth.map((item, index) => (
            <li
              onClick={onOpenTodo}
              data-date={`${item.day}.${currentMonth}.${currentYear}`}
              className={`calendar__day calendar__day_event
               ${item.isHoliday === 1 && 'calendar__day_holiday'} ${
                item.isHoliday === 2 && 'calendar__day_inactive'
              }`}
              key={index}>
              {item.day}
            </li>
          ))}
        </ul>
      </div>

      <p className="calendar__week-deals">
        Желаете посмотреть планы на неделю?{' '}
        <span onClick={onOpenWeekTasks} className="calendar__see-more">
          Посмореть
        </span>
      </p>
    </section>
  );
};

export default Calendar;
