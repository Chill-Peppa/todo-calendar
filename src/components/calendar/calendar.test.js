import React from 'react';
import renderer from 'react-test-renderer';
import Calendar from './calendar';
import { render, screen, fireEvent } from '@testing-library/react';
import { checkDate } from './calendar';
import '@testing-library/jest-dom/extend-expect';

const onOpenTodo = jest.fn();
const onOpenWeekTasks = jest.fn();

it('should correctly render Calendar', () => {
  const tree = renderer
    .create(
      <Calendar onOpenTodo={onOpenTodo} onOpenWeekTasks={onOpenWeekTasks} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should open modal with week tasks', () => {
  render(
    <Calendar onOpenTodo={onOpenTodo} onOpenWeekTasks={onOpenWeekTasks} />,
  );

  const openWeekTasksButton = screen.getByTestId('open-button');
  fireEvent.click(openWeekTasksButton);

  expect(onOpenWeekTasks).toHaveBeenCalled();
});

it('checkDate should return true if the date is found in the savedTodoList', () => {
  const date = '2022-01-01';
  localStorage.setItem(
    'todoList',
    JSON.stringify([
      { id: 1, todo: 'Do something', isDone: false, date: '2022-01-01' },
    ]),
  );

  const result = checkDate(date);

  expect(result).toBe(true);
});

it('checkDate should return false if the date is not found in the savedTodoList', () => {
  const date = '2022-01-02';
  localStorage.setItem(
    'todoList',
    JSON.stringify([
      { id: 1, todo: 'Do something', isDone: false, date: '2022-01-01' },
    ]),
  );

  const result = checkDate(date);

  expect(result).toBe(false);
});

it('onNextButtonClick updates currentMonth and currentYear correctly', () => {
  render(
    <Calendar onOpenTodo={onOpenTodo} onOpenWeekTasks={onOpenWeekTasks} />,
  );
  const nextButton = screen.getByTestId('next-button');
  fireEvent.click(nextButton);

  expect(screen.getByTestId('calendar-title')).toHaveTextContent('Июль 2024');
});

it('onPrevButtonClick updates currentMonth and currentYear correctly', () => {
  render(
    <Calendar onOpenTodo={onOpenTodo} onOpenWeekTasks={onOpenWeekTasks} />,
  );
  const nextButton = screen.getByTestId('prev-button');
  fireEvent.click(nextButton);

  expect(screen.getByTestId('calendar-title')).toHaveTextContent('Май 2024');
});
