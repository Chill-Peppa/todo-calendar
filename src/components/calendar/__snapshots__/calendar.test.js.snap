// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Calendar 1`] = `
<section
  className="calendar"
>
  <div
    className="calendar__header"
  >
    <h1
      className="calendar__title"
      data-testid="calendar-title"
    >
      Июнь 2024
    </h1>
    <div
      className="calendar__button-container"
    >
      <button
        className="calendar__button"
        onClick={[Function]}
      >
        <img
          alt="left arrow"
          className="calendar__prev"
          data-testid="prev-button"
          src="left-arrow.svg"
        />
      </button>
      <button
        className="calendar__button"
        onClick={[Function]}
      >
        <img
          alt="right arrow"
          className="calendar__next"
          data-testid="next-button"
          src="right-arrow.svg"
        />
      </button>
    </div>
  </div>
  <div
    className="calendar__main"
  >
    <ul
      className="calendar__weekdays"
    >
      <li
        className="calendar__weekday"
      >
        Пн
      </li>
      <li
        className="calendar__weekday"
      >
        Вт
      </li>
      <li
        className="calendar__weekday"
      >
        Ср
      </li>
      <li
        className="calendar__weekday"
      >
        Чт
      </li>
      <li
        className="calendar__weekday"
      >
        Пт
      </li>
      <li
        className="calendar__weekday"
      >
        Сб
      </li>
      <li
        className="calendar__weekday"
      >
        Вс
      </li>
    </ul>
    <ul
      className="calendar__days"
    />
  </div>
  <p
    className="calendar__week-deals"
  >
    Желаете посмотреть планы на неделю?
     
    <span
      className="calendar__see-more"
      data-testid="open-button"
      onClick={[MockFunction]}
    >
      Посмореть
    </span>
  </p>
</section>
`;
